data-science-onramp/data-ingestion/setup.py

```diff
@@ -0,0 +1,161 @@
+        if col_value is None:
+            return col_value
+        elif allow_none:
+            return random_select([None, proc_func(col_value)],
+                                 [0.05, 0.95])
+        else:
+            return proc_func(col_value)
+    return udf
+
+
+def id(x):
+    return x
+
+
+def write_to_bigquery(spark, df):
+    '''Write a dataframe to BigQuery'''
+
+    # Create BigQuery Dataset
+    client = bigquery.Client()
+    dataset_id = f'{client.project}.new_york_citibike_trips'
+    dataset = bigquery.Dataset(dataset_id)
+    dataset.location = "US"
+    dataset = client.create_dataset(dataset)
+
+    # Saving the data to BigQuery
+    spark.conf.set('temporaryGcsBucket', BUCKET_NAME)
+
+    df.write.format('bigquery') \
+        .option('table', dataset_id + ".RAW_DATA") \
+        .save()
+
+
+def main():
+    # Create a SparkSession under the name "setup". Viewable via the Spark UI
+    spark = SparkSession.builder.appName("setup").getOrCreate()
+
+    upload = True  # Whether to upload data to BigQuery
+
+    # Check whether or not results should be uploaded
+    if len(sys.argv) > 1:
+        upload = False
+        print("Not uploading results to BigQuery")
+    else:
+        print("Results will be uploaded to BigQuery")
+
+    # Check if table exists
+    try:
+        df = spark.read.format('bigquery').option('table', TABLE).load()
+    except Py4JJavaError:
+        print(f"{TABLE} does not exist. ")
+        sys.exit(0)
+
+    # Declare data transformations for each column in dataframe
+    udfs = [
+        (dirty_data(trip_duration, True), StringType()),  # tripduration
+        (dirty_data(id, True), StringType()),  # starttime
+        (dirty_data(id, True), StringType()),  # stoptime
+        (id, IntegerType()),  # start_station_id
+        (dirty_data(station_name, False), StringType()),  # start_station_name
+        (dirty_data(convert_angle, True), StringType()),  # start_station_latitude
+        (dirty_data(convert_angle, True), StringType()),  # start_station_longitude
+        (id, IntegerType()),  # end_station_id
+        (dirty_data(station_name, False), StringType()),  # end_station_name
+        (dirty_data(convert_angle, True), StringType()),  # end_station_latitude
+        (dirty_data(convert_angle, True), StringType()),  # end_station_longitude
+        (id, IntegerType()),  # bikeid
+        (dirty_data(user_type, False), StringType()),  # usertype
+        (id, IntegerType()),  # birth_year
+        (dirty_data(gender, False), StringType()),  # gender
+        (id, StringType()),  # customer_plan
+    ]
+
+    # Apply dirty transformations to df
+    names = df.schema.names
+    new_df = df.select(*[UserDefinedFunction(*udf)(column).alias(name)
+                         for udf, column, name in zip(udfs, df.columns, names)])
+
+    new_df.sample(False, 0.0001, seed=50).show(n=100)
+
+    # Duplicate about 0.01% of the rows
+    dup_df = new_df.sample(True, 0.0001, seed=42)
```

Comment:
There's no intention, we just thought it might be better for this to be deterministic. We'll remove them in the next revision since there is no reason to hard-code them.